In [1]:
import time
from dask.distributed import Client
import joblib

from CIFAR10_CNN import cifar10_cnn
from CIFAR10_CNN_Improved import cifar10_cnn_improved
from CIFAR10_CNN_Distributed import cifar10_cnn_distributed

In [2]:
client = Client(processes=False)             # create local cluster
# client = Client("scheduler-address:8786")  # or connect to remote cluster

# Dask parallel
print('Starting distributed model training...')
start = time.time()
with joblib.parallel_backend('dask'):
    model_1_dist = cifar10_cnn_distributed('cnn_model_1epochs_dist.h5')
    model_10_dist = cifar10_cnn_distributed('cnn_model_10epochs_dist.h5')
    model_50_dist = cifar10_cnn_distributed('cnn_model_50epochs_dist.h5')
    #model_100_dist = cifar10_cnn_distributed('cnn_model_100epochs_dist.h5')
    
    print('Starting distributed model training (epochs=1)...')
    start = time.time()
    model_1_dist.run_test_harness(epochs=1, batch_size=64)
    end = time.time()
    print('No dask completion time (epochs=1): ', end - start, 'seconds')
    
    print('Starting distributed model training (epochs=10)...')
    start = time.time()
    model_10_dist.run_test_harness(epochs=10, batch_size=64)
    end = time.time()
    print('No dask completion time (epochs=10): ', end - start, 'seconds')
    
    print('Starting distributed model training (epochs=50)...')
    start = time.time()
    #model_50_dist.run_test_harness(epochs=50, batch_size=64)
    end = time.time()
    print('No dask completion time (epochs=50): ', end - start, 'seconds')
    
    print('Starting distributed model training (epochs=100)...')
    start = time.time()
    #model_100_dist.run_test_harness(epochs=100, batch_size=64)
    end = time.time()
    print('No dask completion time (epochs=100): ', end - start, 'seconds')
    
print('Distributed models done')

#No dask completion time (epochs=10):  2111.384060382843 seconds
#Starting model training (epochs=50)...
#No dask completion time (epochs=50):  8861.92734336853 seconds
#Starting model training (epochs=100)...
#Starting distributed model training (epochs=10)...
#No dask completion time (epochs=10):  1963.2309262752533 seconds

Starting distributed model training...
Starting distributed model training (epochs=1)...


distributed.worker - WARNING - Compute Failed
Function:  _score
args:      ((KerasClassifier(
	model=<function define_model_improved at 0x0000029038B03DC8>
	build_fn=None
	warm_start=False
	random_state=None
	optimizer=rmsprop
	loss=None
	metrics=None
	batch_size=64
	validation_batch_size=None
	verbose=False
	callbacks=None
	validation_split=0.0
	shuffle=True
	run_eagerly=False
	epochs=1
	learning_rate=0.07269712277405341
	momentum=0.11706618702172067
	class_weight=None
), {'model_id': 3, 'params': {'batch_size': 64, 'epochs': 1, 'learning_rate': 0.07269712277405341, 'momentum': 0.11706618702172067}, 'partial_fit_calls': 1, 'partial_fit_time': 741.5639660358429}), array([[[[0.5176471 , 0.41568628, 0.18431373],
         [0.627451  , 0.5254902 , 0.25882354],
         [0.6627451 , 0.63529414, 0.22352941],
         ...,
         [0.7176471 , 0.627451  , 0.3137255 ],
         [0.6862745 , 0.6       , 0.27450982],
         [0.69411767, 0.60784316, 0.28235295]],

        [[0.5686275 , 0.46274

TypeError: 'NoneType' object is not callable

distributed.worker - WARNING - Compute Failed
Function:  _score
args:      ((KerasClassifier(
	model=<function define_model_improved at 0x0000029038B03DC8>
	build_fn=None
	warm_start=False
	random_state=None
	optimizer=rmsprop
	loss=None
	metrics=None
	batch_size=64
	validation_batch_size=None
	verbose=False
	callbacks=None
	validation_split=0.0
	shuffle=True
	run_eagerly=False
	epochs=1
	learning_rate=0.020195367850306834
	momentum=0.8388640303632403
	class_weight=None
), {'model_id': 2, 'params': {'batch_size': 64, 'epochs': 1, 'learning_rate': 0.020195367850306834, 'momentum': 0.8388640303632403}, 'partial_fit_calls': 1, 'partial_fit_time': 745.8535649776459}), array([[[[0.49019608, 0.65882355, 0.8666667 ],
         [0.49411765, 0.65882355, 0.8666667 ],
         [0.49803922, 0.6666667 , 0.87058824],
         ...,
         [0.50980395, 0.67058825, 0.8784314 ],
         [0.5058824 , 0.67058825, 0.8784314 ],
         [0.49803922, 0.68235296, 0.8784314 ]],

        [[0.49803922, 0.67058

INFO:tensorflow:Assets written to: C:\Users\megal\AppData\Local\Temp\tmp1j8v6gwd\assets


In [2]:
model_event_1 = cifar10_cnn_improved('cnn_event_model_1epochs.h5', 'data_img')
model_event_10 = cifar10_cnn_improved('cnn_event_model_10epochs.h5', 'data_img')
model_event_50 = cifar10_cnn_improved('cnn_event_model_50epochs.h5', 'data_img')
model_event_100 = cifar10_cnn_improved('cnn_event_model_100epochs.h5', 'data_img')
#model_dask = cifar10_cnn_dask()

# No dask no parallel event images
print('Starting non-distributed event training...')
# epochs = 1
print('Starting event model training (epochs=1)...')
start = time.time()
model_event_1.run_test_harness(epochs=1, batch_size=64)
end = time.time()
# epochs = 10
print('Starting event model training (epochs=10)...')
start = time.time()
model_event_10.run_test_harness(epochs=10, batch_size=64)
end = time.time()
print('Non-distributed event model completion time (epochs=10): ', end - start, 'seconds')
# epochs = 50
print('Starting event model training (epochs=50)...')
start = time.time()
model_event_50.run_test_harness(epochs=50, batch_size=64)
end = time.time()
print('Non-distributed event model completion time (epochs=50): ', end - start, 'seconds')
# epochs = 100
print('Starting event model training (epochs=100)...')
start = time.time()
model_event_50.run_test_harness(epochs=100, batch_size=64)
end = time.time()
print('Non-distributed event model completion time (epochs=100): ', end - start, 'seconds')

Starting non-distributed event training...
Starting event model training (epochs=1)...
Found 65000 files belonging to 10 classes.
Using 58500 files for training.
Found 65000 files belonging to 10 classes.
Using 6500 files for validation.


C:\Users\megal\Coursework\CS573\Project\CIFAR10_CNN.py:72: VisibleDeprecationWarning: Creating an ndarray from ragged nested sequences (which is a list-or-tuple of lists-or-tuples-or ndarrays with different lengths or shapes) is deprecated. If you meant to do this, you must specify 'dtype=object' when creating the ndarray.
  trainX = np.array(list(map(lambda x: x[0], train_numpy)))
C:\Users\megal\Coursework\CS573\Project\CIFAR10_CNN.py:74: VisibleDeprecationWarning: Creating an ndarray from ragged nested sequences (which is a list-or-tuple of lists-or-tuples-or ndarrays with different lengths or shapes) is deprecated. If you meant to do this, you must specify 'dtype=object' when creating the ndarray.
  trainY = np.array(list(map(lambda x: x[1], train_numpy)))
C:\Users\megal\Coursework\CS573\Project\CIFAR10_CNN.py:77: VisibleDeprecationWarning: Creating an ndarray from ragged nested sequences (which is a list-or-tuple of lists-or-tuples-or ndarrays with different lengths or shapes) is d

> 13.646
Starting event model training (epochs=10)...
Found 65000 files belonging to 10 classes.
Using 58500 files for training.
Found 65000 files belonging to 10 classes.
Using 6500 files for validation.
> 13.769
Non-distributed event model completion time (epochs=10):  2691.480133533478 seconds
Starting event model training (epochs=50)...
Found 65000 files belonging to 10 classes.
Using 58500 files for training.
Found 65000 files belonging to 10 classes.
Using 6500 files for validation.
> 14.631
Non-distributed event model completion time (epochs=50):  13552.021950244904 seconds
Starting event model training (epochs=100)...
Found 65000 files belonging to 10 classes.
Using 58500 files for training.
Found 65000 files belonging to 10 classes.
Using 6500 files for validation.
> 14.631
Non-distributed event model completion time (epochs=100):  66.64784717559814 seconds


Starting non-distributed event training...
Starting event model training (epochs=1)...
Found 65000 files belonging to 10 classes.
Using 58500 files for training.
Found 65000 files belonging to 10 classes.
Using 6500 files for validation.
> 13.646
Starting event model training (epochs=10)...
Found 65000 files belonging to 10 classes.
Using 58500 files for training.
Found 65000 files belonging to 10 classes.
Using 6500 files for validation.
> 13.769
Non-distributed event model completion time (epochs=10):  2691.480133533478 seconds
Starting event model training (epochs=50)...
Found 65000 files belonging to 10 classes.
Using 58500 files for training.
Found 65000 files belonging to 10 classes.
Using 6500 files for validation.
> 14.631
Non-distributed event model completion time (epochs=50):  13552.021950244904 seconds
Starting event model training (epochs=100)...
Found 65000 files belonging to 10 classes.
Using 58500 files for training.
Found 65000 files belonging to 10 classes.
Using 6500 files for validation.

In [3]:
import tensorflow_datasets as tfds
from keras.preprocessing import image
from tensorflow.data import AUTOTUNE
from tensorflow.keras.layers import Rescaling
from tensorflow.keras.preprocessing import image_dataset_from_directory
import numpy as np

train = image_dataset_from_directory(
    'data_img',
    label_mode='categorical',
    validation_split=0.1,
    subset="training",
    seed=0,
    color_mode="grayscale",
    batch_size=32,
    image_size=(32, 32)
)
test = image_dataset_from_directory(
    'data_img',
    label_mode='categorical',
    validation_split=0.1,
    subset="validation",
    seed=0,
    color_mode="grayscale",
    batch_size=32,
    image_size=(32, 32)
)

#normalization_layer = Rescaling(1./255)
#normalized_ds = train.map(lambda x, y: (normalization_layer(x), y))
#image_batch, labels_batch = next(iter(normalized_ds))
#first_image = image_batch[0]
# Notice the pixel values are now in `[0,1]`.
#print(np.min(first_image), np.max(first_image))

#train = train.cache().prefetch(buffer_size=AUTOTUNE)
#test = test.cache().prefetch(buffer_size=AUTOTUNE)

train_numpy = tfds.as_numpy(train)  # Convert `tf.data.Dataset` to Python generator

X_train = np.array(list(map(lambda x: x[0], train_numpy)))
y_train = np.array(list(map(lambda x: x[1], train_numpy)))

#print(X_train.shape)
#print(X_train[0].shape)

y_train = np.concatenate(y_train[:])
#y_train = y_train.reshape((len(y_train), 1))
print(y_train.shape)
print(y_train[0].shape)
print(y_train)

#X_train = np.concatenate(X_train[:])

#print(X_train.shape)
#print(X_train[0].shape)
#train_norm = X_train.astype('float32')
#for ex in ds_numpy:
    # `{'image': np.array(shape=(28, 28, 1)), 'labels': np.array(shape=())}`
#    print(ex[1])
#    break

Found 10000 files belonging to 10 classes.
Using 9000 files for training.
Found 10000 files belonging to 10 classes.
Using 1000 files for validation.
(9000, 10)
(10,)
[[0. 0. 0. ... 1. 0. 0.]
 [0. 0. 0. ... 0. 0. 0.]
 [1. 0. 0. ... 0. 0. 0.]
 ...
 [0. 0. 0. ... 1. 0. 0.]
 [0. 0. 0. ... 1. 0. 0.]
 [0. 0. 0. ... 0. 0. 0.]]


In [2]:
from keras.datasets import cifar10
from tensorflow.keras.utils import to_categorical

(trainX, trainY), (testX, testY) = cifar10.load_data()
print(trainY.shape)
trainY = to_categorical(trainY)
print(trainY.shape)
print(trainY)

(50000, 1)
(50000, 10)
[[0. 0. 0. ... 0. 0. 0.]
 [0. 0. 0. ... 0. 0. 1.]
 [0. 0. 0. ... 0. 0. 1.]
 ...
 [0. 0. 0. ... 0. 0. 1.]
 [0. 1. 0. ... 0. 0. 0.]
 [0. 1. 0. ... 0. 0. 0.]]
